lib/charms/prometheus_k8s/v1/prometheus.py

```diff
@@ -317,99 +314,71 @@ def __init__(self, charm, name, consumes, multi=False):
-                `multi` mode.
+        When a consumer charm is related to a Prometheus provider, the
+        consumer sets metadata related to its own scrape
+        configutation.  This metadata is set using Juju application
+        data.  In addition each of the consumer units also sets its own
+        host address in Juju unit relation data.
         """
-        if rel_id is None:
-            rel_id = self.relation_id
+        event.relation.data[self._charm.unit]["prometheus_scrape_host"] = str(
+            self._charm.model.get_binding(event.relation).network.bind_address
+        )
 
-        targets = self._stored.targets.get(rel_id, [])
-        if address in targets:
+        if not self._charm.unit.is_leader():
             return
 
-        target = address + ":" + str(port)
-        targets.append(target)
-        self._update_targets(targets, rel_id)
-
-    def remove_endpoint(self, address, port=80, rel_id=None):
-        """Remove an endpoint from the list of Prometheus scrape targets.
-        Args:
-            address: a string host address (usually IP) of the endpoint that
-                that must be excluded from being monitored by Prometheus.
-            port: an optional (default 80) integer providing the port
-                on which the scrapped endpoint exposes its Prometheus
-                metrics.
-            rel_id: an optional integer providing the relation ID for
-                the related Prometheus monitoring service
-                provider. This is only necessary if the
-                `PrometheusConsumer` has been instantiated in
-                `multi` mode.
-        """
-        if rel_id is None:
-            rel_id = self.relation_id
-
-        targets = self._stored.targets.get(rel_id, [])
-        target = address + ":" + str(port)
-        if target not in targets:
-            return
+        event.relation.data[self._charm.app]["prometheus_scrape_metadata"] = json.dumps(
+            self._scrape_metadata
+        )
 
-        targets.remove(target)
-        self._update_targets(targets, rel_id)
+    def _set_unit_ip(self, event):
+        """Set unit host address
 
-    @property
-    def endpoints(self, rel_id=None):
-        """Returns a list of Prometheus scrape target endpoints.
-        Args:
-            rel_id: an optional integer providing the relation ID for
-                the related Prometheus monitoring service
-                provider. This is only necessary if the
-                `PrometheusConsumer` has been instantiated in
-                `multi` mode.
+        Each time a consumer charm container is restarted it updates its own
+        host address in the unit relation data for the Prometheus provider.
         """
+        for relation in self._charm.model.relations[self._relation_name]:
+            relation.data[self._charm.unit]["prometheus_scrape_host"] = str(
+                self._charm.model.get_binding(relation).network.bind_address
+            )
```

Comment:
It might be worth reaching out to someone from the Juju team familiar with networking but I worry a bit that, particularly for CMR, there's a chance that the address that the Prometheus side needs to connect to could be different than the address the target side should bind to. My understanding is that this edge case is why `ingress-address` and `bind-address` are two separate fields.

Reply:
My understanding was `ingress-address` was going to be deprecated. Thanks again for pointing this out. I will get it clarified by Juju/Operator teams.

Reply:
Other way around -- `private-address` is deprecated, leaving `ingress-address` only. See https://github.com/juju/juju/blob/4ae860b247b4f3e821aa333ff4bf85af1d20b9d0/apiserver/facades/agent/uniter/uniter.go#L1389.

In any case, my hope for the concern raised by @johnsca is that Juju/OF will be "smart" enough to give you the right address in context for the relation

Reply:
So what is the conclusion here ? Should I switch the bind address to ingress address ? . I have created a ticket for this here, https://github.com/canonical/prometheus-operator/issues/72 just in case this question does not have a definitive answer as yet.

Reply:
We should ask the Juju team what their plans for this are. My guess is that this is more or less deferred while they figure out cross-model relations and which address should be selected. I'd leave it as-is, if it were me.